annnufan/assignment4/islands_counter.cpp

```diff
@@ -0,0 +1,59 @@
+#include <vector>
+#include <utility>
+
+class islands_counter {
+	std::vector<std::vector<bool>> island_map;
+	std::vector<std::pair<int, int>> moves;
+	int n, m;
+	int islands;
+	bool is_count;
+
+	void dfs(int i, int j) {
+		island_map[i][j] = false;
+
+		for (const auto& move : moves) {
+			int i1 = move.first + i;
+			int j1 = move.second + j;
+			if (0 <= i1 && i1 < n && 0 <= j1 && j1 < m && island_map[i1][j1]) {
+				dfs(i1, j1);
+			}
+		}
+	}
+
+public:
+	islands_counter(std::vector<std::vector<bool>> i_m) : island_map(i_m) {

```

Comment:
You pass here `i_m` by value, which means the entire map is copied, twice:
1. Once when passing it into the function (the parameter is copy-constructed)
2. Once when copying the parameter into the `island_map` field.

You can avoid the 1st copy by passing it by const reference:
`islands_counter(const std::vector<std::vector<bool>>& i_m)`
Here and in the functions below.

Another option is to "move" keep the parameter by value, but "move" it into the field:
```
islands_counter(std::vector<std::vector<bool>> i_m) : island_map(std::move(i_m)) ...
```
This uses C++11 move semantics. This is out of our scope so I don't want to get into it, just know that it exists if you see something like this in the future. :)